slack-lint-checks/src/main/java/slack/lint/util/SlackJavaEvaluator.kt

```diff
@@ -0,0 +1,229 @@
+
+  override fun getAllAnnotations(
+    owner: PsiModifierListOwner,
+    inHierarchy: Boolean
+  ): Array<PsiAnnotation> = delegate.getAllAnnotations(owner, inHierarchy)
+
+  override fun getAllAnnotations(owner: UAnnotated, inHierarchy: Boolean): List<UAnnotation> =
+    delegate.getAllAnnotations(owner, inHierarchy)
+
+  override fun getAnnotation(
+    listOwner: PsiModifierListOwner?,
+    vararg annotationNames: String
+  ): UAnnotation? = delegate.getAnnotation(listOwner, *annotationNames)
+
+  override fun getAnnotationInHierarchy(
+    listOwner: PsiModifierListOwner,
+    vararg annotationNames: String
+  ): UAnnotation? = delegate.getAnnotationInHierarchy(listOwner, *annotationNames)
+
+  override fun getAnnotations(
+    owner: PsiModifierListOwner?,
+    inHierarchy: Boolean,
+    parent: UElement?
+  ): List<UAnnotation> = delegate.getAnnotations(owner, inHierarchy, parent)
+
+  override fun getClassType(psiClass: PsiClass?): PsiClassType? = delegate.getClassType(psiClass)
+
+  override fun getPackage(node: PsiElement): PsiPackage? = delegate.getPackage(node)
+
+  override fun getPackage(node: UElement): PsiPackage? = delegate.getPackage(node)
+
+  override fun getTypeClass(psiType: PsiType?): PsiClass? = delegate.getTypeClass(psiType)
+
+  override fun implementsInterface(cls: PsiClass, interfaceName: String, strict: Boolean): Boolean =
+    delegate.implementsInterface(cls, interfaceName, strict)
+  // endregion
+
+  override fun hasModifier(owner: PsiModifierListOwner?, keyword: KtModifierKeywordToken): Boolean {
+    val superValue = super.hasModifier(owner, keyword)
+    // If it's not a compiled element or not a PsiClass, trust the super value and move on
+    if (!checkMetadata || owner !is PsiCompiledElement || owner !is PsiClass) {
+      return superValue
+    }
+
+    // We're working with an externally compiled element and it's a PsiClass, so we can do more
+    // thorough checks here.
+    KOTLIN_CLASS_MODIFIER_TOKENS[keyword]?.let { (flag, applicableClassKinds) ->
+      owner.findContaining(UClass::class.java)?.let { cls ->
+        // Only parse if the target class kind is applicable to the token we're checking. For
+        // example - when checking `data` tokens, they're not applicable to interfaces or enums.
+        if (cls.classKind in applicableClassKinds) {
+          cls.getOrParseMetadata()?.let { kmClass ->
+            return flag(kmClass.flags)
+          }
+        }
+      }
+    }
+
+    return superValue
+  }
+
+  private fun UAnnotated.getOrParseMetadata(): KmClass? {
+    val cls =
+      when (this) {
+        is UClass -> this
+        else -> return null // Only classes are supported right now
+      }
+    return cachedClasses
+      // Don't use getOrPut. Kotlin's extension may still invoke the body and we don't want that
+      .computeIfAbsent(qualifiedName!!) { key ->
+        val annotation =
+          cls.findAnnotation("kotlin.Metadata") ?: return@computeIfAbsent Optional.empty()
```

Comment:
Same here. It seems like you should be able to `return@computeIfAbsent null` here

Reply:
I dislike returns with labels and do my best to avoid them when possible (I think they negatively affect readability). It isn't a huge deal (I understand not everyone feels this way), but did you consider something like this:

```kolin
var metadata = cachedClasses[qualifiedName!!]
if (metadata == null) {
  val annotation = cls.findAnnotation("kotlin.Metadata") ?: return null
  val (durationMillis, result) = measureTimeMillisWithResult { annotation.parseMetadata(key) }
  slackLintLog("Took ${durationMillis}ms to parse metadata for $key.")
  metadata = result
}

return metadata
```

Reply:
See above for the nullability issue

For `computeIfAbsent` - we _have_ to use it because it guarantees the currency safety and that the block is only computed once. The code in your snippet is actually more or less what kotlin's `ConcurrentHashmap.getOrPut` does, which is actually racy and could call the computation block more than once. See the comment above this call too :P

Reply:
Ahh yes 👍 . I read the docs for `getOrPut` but didn't think to look at the docs for `computeIfAbsent`